### ISSUE TITLE
C assumes ordered DeviceUIDs

### ISSUE THREAD

**Issue by jakebeal** [Protelis/Protelis#130]
Our implementation of C in protelis:coord:accumulation assumes that DeviceUIDs are numbers.

In particular, the call to getParent uses getDeviceUID():
`
getParent(potential, (v) -> { v.getDeviceUID() })) == self.getDeviceUID()) 
`
and then getParentExtended uses it for a tiebreaker in minHood over tuples:
`        (v) -> { minHood(v) },`

When DeviceUIDs are not comparable (e.g., right now I am using string UIDs), this causes C to fail, because parents cannot be correctly identified.

I see two possible ways to resolve this:

1. Use a stable random nonce rather than a DeviceUID in C (this is my current kludge)
2. Require that DeviceUIDs implement Comparable and make sure minHood can handle arbitrary Comparable objects.

I think that my preference is for the first, but it could go either way.

**Comment by w4bo** [comment]
However, in this case
```
if(trueAfter(X)) {
    C(...)
} else {
    C(...)
}
```
will return to 2 different fields over time. Are we ok with this? Actually, this is the same for `S`

**Comment by w4bo** [comment]
Also, I was trying to reproduce the bug. Even if I change `C` to
```
def C(potential, reduce, local, null) {
   ...
   mux (nbr(getParent(potential, (v) -> { "abc" + v.getDeviceUID().toString() })) == "abc" + self.getDeviceUID().toString())
   ...
}
``` 
All the tests in protelis-lang succeed.

This is the stack of functions:
```
def C(potential, reduce, local, null) {
   ...
   mux (nbr(getParent(potential, v -> { v.getDeviceUID() })) == self.getDeviceUID())
   ...
}
def getParent(potential, f) {
    getParentExtended(potential, f, value -> { NaN }, NaN).get(0)
}
def getParentExtended(potential, f, g, local) {
    getParentsExtended(potential, v -> { minHood(v) }, f, g, local, local)
}
def getParentsExtended(potential, condition, f, g, local, default) {
    mux (condition.apply(nbr(potential)) < potential) {
        let c = condition.apply(nbr([potential, f.apply(self), g.apply(local)]));
        ...
    } ...
}
```
`v -> { minHood(v) }`  is applied first to `nbr(potential)`, and then to `nbr([potential, v -> { v.getDeviceUID() }, v -> { NaN }])`

**Comment by jakebeal** [comment]
Java Strings also implement Comparable, so your modified test will still succeed.  The problem is that DeviceUID does *not* extend Comparable.

As for the potential problem of two instances of C tiebreaking differently: yes, that would be a side effect, and is a possible argument for solution #2.  Danilo, do you have an opinion?

**Comment by DanySK** [comment]
I personally believe extending `Comparable` might be problematic. There are two reasons:
* `Comparable` is a generic interface, so we would need to actually extend something like `Comparable<DeviceUID>`. But we don't take any stance about what a `DeviceUID` is, and that would mean that writing meaningful `compareTo` methods would be very error-prone (instanceof + cast or exception, or fallback to lexicographic comparison).
* `DeviceUID` needs to be comparable just because of the tie-breaking strategy we picked in an algorithm, when other strategies are available too (e.g. `self.getDeviceUID().toString()` and use that for comparison).

I would personally suggest to document the interface and suggest that implementations should also implement `Comparable` of itself if the back-end is intended to be able to support all the library algorithms. We could also provide a couple of implementations using UUIDs, Strings and longs.

Alternatively, we could make `DeviceUID` generic, something like: `DeviceUID<T> extends `Comparable<T>`, and have a method to retrieve T, but I find this rather ugly.

**Comment by jakebeal** [comment]
You raise some good points, Danilo, and I've been struggling with this question.  

I do think that UIDs have proven to be very important for tie-breaking, and that an expensive kludge strategy like `self.getDeviceUID().toString()` are not a good way to doing an operation that is likely to be so frequent.

Accordingly, I think that we should adopt your suggestion to document that all implementations are recommended to implement Comparable, and provide UUID, String, and int/long implementations.

I do note that if we could test whether a DeviceUID is Comparable, then if it's not one could fall back to the string method.  But I can't think of any reasonable way to insert this.  We we do need, however, is library functions like C to error on non-comparability.  Right now, it fails silently, which I think is due to problems in how Tuple comparison is implemented.

**Comment by jakebeal** [comment]
Fixed by #136.
